staging/src/k8s.io/apiserver/pkg/server/serve.go

```diff
@@ -224,22 +238,216 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+	return nil
+}
+
+// ipBasedLimitHandler accepts at most n simultaneous connections from a
+// given IP address. We also limit the total simultaneous connections.
+// None of this is applied to localhost/loopback connections.
+//
+// Based on Go 1.7.4 net/netutil/listen.go
+// TODO: (cheftako) : IPv6, CIDR Range rather than individual IP Addresses?
+// Considered using atomic for per IP limit value,
+// but problem initing map value safely
+
+type ipBasedLimit struct {
+	// The maximum concurent connections an IP is allowed to have open
+	maxPerIP int
+
+	// The total maximum concurrent connections allowed open except those from localhost
+	max int
+
+	// Lock used to ensure the limits and total are updated atomicaly
+	mutex sync.Mutex
+
+	// Map in which we map the serialized ip to the current count
+	limits map[string]int
+
+	// Map in which we retain the lookup from connection to decrementIP call
+	decrementers map[string]func()
+
+	// Current total count of the connections in existance except those from localhost
+	total int
+}
+
+func (i *ipBasedLimit) incrementIP(conn, sourceIP string) error {
+	if i.max <= 0 {
+		return nil
+	}
+	ip := net.ParseIP(sourceIP)
+	if ip == nil {
+		glog.Infof("Received request from client with invalid ip %s", sourceIP)
+		return IPLimitExceededError{ip: sourceIP, reason: "unparseable ip address"}
+	}
+	if ip.IsLoopback() {
+		return nil
+	}
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+	current := i.limits[sourceIP]
+	if current >= i.maxPerIP {
+		return IPLimitExceededError{ip: sourceIP, reason: fmt.Sprintf(" exceeded per IP limit of %d", i.maxPerIP)}
+	}
+	if i.total >= i.max {
+		return IPLimitExceededError{ip: sourceIP, reason: fmt.Sprintf(" exceeded overall limit of %d", i.max)}
+	}
+	i.total++
+	i.limits[sourceIP] = current + 1
+	var releaseOnce sync.Once
+	i.decrementers[conn] = func() {
+		releaseOnce.Do(func() { i.decrementIP(conn, sourceIP) })
+	}
+	return nil
+}
+
+func (i *ipBasedLimit) decrementIP(conn, sourceIP string) {
+	if i.max <= 0 {
+		return
+	}
+	ip := net.ParseIP(sourceIP)
+	if ip == nil || ip.IsLoopback() {
+		return
+	}
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+	i.total--
+	current, ok := i.limits[sourceIP]
+	if !ok {
+		glog.Errorf("Attempt to decrement limit for not connected IP %s", sourceIP)
+	}
+	if current > 1 {
+		i.limits[sourceIP] = current - 1
+	} else {
+		delete(i.limits, sourceIP)
+	}
+	delete(i.decrementers, conn)
 }
 
-func (ln tcpKeepAliveListener) Accept() (net.Conn, error) {
-	tc, err := ln.AcceptTCP()
+// Using a custom marshal JSON method for several reasons
+// A) It allows us to minimize the scope of the mutex which needs to be held
+// B) It allows the fields to remain private
+// C) It allows us to marshal 'decrementers' as a list of its keys
+// This is only meant for use in the debug API.
+func (i *ipBasedLimit) MarshalJSON() ([]byte, error) {
+	buffer := bytes.NewBufferString("{")
+	buffer.WriteString(fmt.Sprintf("\"MaxPerIP\": %d, ", i.maxPerIP))
```

Comment:
Also, backticks will make the quoting more readable. (but also mean that \n wouldn't be a newline, if that matters.)